build-index/tagserver/server.go

```diff
@@ -405,3 +451,68 @@ func (s *Server) replicateTag(tag string, d core.Digest, deps core.DigestList) e
 	}
 	return nil
 }
+
+func buildPaginationOptions(u *url.URL) (*[]backend.ListOption, error) {
+	var opts []backend.ListOption
+	q := u.Query()
+	for k, v := range q {
+		if len(v) != 1 {
+			return nil, handler.Errorf(
+				"invalid query %s:%s", k, v).Status(http.StatusBadRequest)
+		}
+		if k == limitQ {
+			limitCount, err := strconv.Atoi(v[0])
+			if err != nil {
+				return nil, handler.Errorf(
+					"invalid limit %s: %s", v, err).Status(http.StatusBadRequest)
+			}
+			if limitCount == 0 {
+				return nil, handler.Errorf(
+					"invalid limit %d", limitCount).Status(http.StatusBadRequest)
+			}
+			opts = append(opts, backend.ListWithMaxKeys(limitCount))
+		} else if k == offsetQ {
+			opts = append(opts, backend.ListWithContinuationToken(v[0]))
+		} else {
+			return nil, handler.Errorf(
+				"invalid query %s", k).Status(http.StatusBadRequest)
+		}
+	}
+	if len(q) > 0 {
```

Comment:
If query exists, we are validating it above. If its validate, it would be related to pagination as of current implementation. But I think enablePagination would make it more readable too. Will fix it.

Reply:
Maybe len(opts) would be good check.